### PR TITLE
Login Module and Instructions module Enhancement

### PR DESCRIPTION
**Pull request by shresthdeepgupta**
added the code for :

-  Student Login 

- Pre Examination Instructions

### CODE REVIEW

**Comment by nidhi-einn** [comment]
Why binary file is committed android/MaterialDesignTest/app/.idea/caches/build_file_checksums.ser?

**Comment by nidhi-einn** [comment]
What is the need to add android/MaterialDesignTest/app/.idea/ repo into our directory

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/network/retrofit/ErrorHandling.kt [hunk 1]
```diff
@@ -35,7 +36,7 @@ fun <T> Flowable<T>.handelNetworkError() =
             is OfflineException ->
                 return@onErrorResumeNext Flowable.error(Exception("check your internet connection"))
             is SocketTimeoutException ->
-                return@onErrorResumeNext Flowable.error(Exception("server not found"))
+                return@onErrorResumeNext Flowable.error(Exception("server not fount"))
```
Comment (nidhi-einn) [inline android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/network/retrofit/ErrorHandling.kt:39]: It should be "Server not Found"

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/ExamInfoPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(
+                    {success->
+                        fetchExamInfo(FetchExamRequest(success.studentCollegeCode,
+                            year, month, dayOfMonth))
+                    },
+                    {
+                        error->
+                        println(error.localizedMessage)
```
Comment (nidhi-einn) [inline android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/ExamInfoPresenter.kt:73]: Put log here

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/ExamInfoPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                    },
+                    {
+                        println("on Complete")
```
Comment (nidhi-einn) [inline android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/ExamInfoPresenter.kt:76]: Put log here

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/InstructionPresenter.kt [hunk 4]
```diff
@@ -16,33 +16,29 @@ class InstructionPresenter(private var view: InstructionsContract.View?):
...
+                    { instructions ->
+                        view!!.showInstructions(instructions!!)
                     },
-                    {err ->
-                        println(err.localizedMessage)
-                        Log.e(TAG, "Error in fetching Instruction from Remote")
-                    }
-
+                    {error -> println(error.toString())},
+                    { println("instructions fetched successfully")}
```
Comment (nidhi-einn) [inline android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/InstructionPresenter.kt:41]: Put log

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/login/LoginPresenter.kt [hunk 5]
```diff
@@ -21,33 +22,32 @@ class LoginPresenter(private var view: LoginContract.View?) :
...
-                                  )
-                                  userRepository.saveStudent(response.token)
+                              { session ->
+                                  Log.d("inside passed Auth", "updating session with email \n ${session.email} \n")
+                                  updateSession(session)
+                                  userRepository.saveStudent(session.token)
+                                  view!!.showLoading(false)
                                   view!!.openMainActivity()
                               },
                               { error -> Log.d("REQUEST FAILED !! ", error.toString()) }
```
Comment (nidhi-einn) [inline android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/login/LoginPresenter.kt:53]: Make it error log

File: android/MaterialDesignTest/app/src/test/java/com/testexample/materialdesigntest/data/network/repository/PreExamInstructionsRemoteRepoTest.kt [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    @After
+    fun tearDown() {
+    }
+
+    @Test
+    fun callInstructionsApi() {
+        instructionRemoteRepoPreExam.callInstructionsApi(token,  "")
+            .subscribe(
+                {success -> println("Instruction is $success")},
+                { error -> println(error.localizedMessage) })
```
Comment (nidhi-einn) [inline android/MaterialDesignTest/app/src/test/java/com/testexample/materialdesigntest/data/network/repository/PreExamInstructionsRemoteRepoTest.kt:38]: Put log

File: android/MaterialDesignTest/app/src/test/java/com/testexample/materialdesigntest/data/network/repository/UserRemoteRepositoryTest.kt [hunk 7]
```diff
@@ -69,9 +65,10 @@ class UserRemoteRepositoryTest {
...
         var getUserRequest = UserRequest("","")
-         repository.authStudent(validStudentLoginRequest)
+         val output = repository.authStudent(validStudentLoginRequest)
              .handelNetworkError()
-             .subscribe(
+        output.test().assertNoErrors()
+        output.subscribe(
                 {success ->
                     getUserRequest = UserRequest(success.token, success.id)
                     println("Auth Response is $success")},
```
Comment (nidhi-einn) [inline android/MaterialDesignTest/app/src/test/java/com/testexample/materialdesigntest/data/network/repository/UserRemoteRepositoryTest.kt:74]: Put log here.

**Comment by nidhi-einn** [comment]
@shresthdeepgupta please address my comments. 

@akanksha-einn and @kanchan-einn I have reviewed this branch code according to my best knowledge. I want one of you guys please review his code.

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/network/retrofit/ErrorHandling.kt (hunk 1 above)
Comment (shresthdeepgupta) [reply android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/network/retrofit/ErrorHandling.kt:39]: i

Comment (shresthdeepgupta) [reply android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/network/retrofit/ErrorHandling.kt:39]: updated it

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/ExamInfoPresenter.kt (hunk 2 above)
Comment (shresthdeepgupta) [reply android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/ExamInfoPresenter.kt:73]: okay

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/ExamInfoPresenter.kt (hunk 3 above)
Comment (shresthdeepgupta) [reply android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/instructions/ExamInfoPresenter.kt:76]: actually logs cant be used in testing

**Comment by shresthdeepgupta** [comment]
> What is the need to add android/MaterialDesignTest/app/.idea/ repo into our directory  

 .idea is used for both project build configuration (dependencies declaration) and project settings (inspections, etc.)

**Comment by shresthdeepgupta** [comment]
> Why binary file is committed android/MaterialDesignTest/app/.idea/caches/build_file_checksums.ser?

I'll check if it necessary or not and then move to ignore if required

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/login/LoginPresenter.kt (hunk 5 above)
Comment (shresthdeepgupta) [reply android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/ui/login/LoginPresenter.kt:53]: okay

File: android/MaterialDesignTest/app/src/test/java/com/testexample/materialdesigntest/data/network/repository/PreExamInstructionsRemoteRepoTest.kt (hunk 6 above)
Comment (shresthdeepgupta) [reply android/MaterialDesignTest/app/src/test/java/com/testexample/materialdesigntest/data/network/repository/PreExamInstructionsRemoteRepoTest.kt:38]: Logging part will be done by @akanksha-einn

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt [hunk 8]
```diff
@@ -20,7 +20,7 @@ class ExaminationRepo(context: Context) : IExaminationRepo {
     private val roomRepo: IExaminationRoomRepo =
         ExaminationRoomRepo(context)
 
-    override fun loadQuestionPaperFromRemote(collegeCode: String, date: String): Single<QuestionPaperComplete> {
+    override fun loadQuestionPaperFromRemote(collegeCode: String, date: String): Single<QuestionPaper> {
         return remoteRepo
             .callApiForQuestionPaper(token, FetchExamRequest(802,2020,5,29))
```
Comment (akanksha-einn) [inline android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt:25]: We should not use hard code value here.

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt [hunk 9]
```diff
@@ -40,20 +39,19 @@ class ExaminationRepo(context: Context) : IExaminationRepo {
...
         roomRepo
-            .addQuestion(questions)
+            .addQuestionForRoom(questions)
     }
 
-    override fun addQuestionPaper(questionPaper: QuestionPaperComplete) {
-        roomRepo
-            .addQuestionPaper(QuestionPaperComplete(QuestionPaper("534","name",50,
-                180,"dada",802,29,5,2020,"10:00 AM",true,1,.25),
-            listOf(Section("534", "section1",52,54, listOf("s54","545","454")))))
```
Comment (akanksha-einn) [inline android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt:56]: We should not use hard code value in function call here.

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt [hunk 10]
```diff
@@ -40,20 +39,19 @@ class ExaminationRepo(context: Context) : IExaminationRepo {
...
+    override fun addQuestionPaper(questionPaper: QuestionPaper) {
+//        roomRepo
+//            .addQuestionPaper(QuestionPaper(QuestionPaper("534","name",50,
+//                180,"dada",802,29,5,2020,"10:00 AM",true,1,.25),
+//            listOf(Section("534", "section1",52,54, listOf("s54","545","454")))))
     }
```
Comment (akanksha-einn) [inline android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt:55]: Please remove comment

Comment (shresthdeepgupta) [reply android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt:55]: okay

File: android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt (hunk 8 above)
Comment (shresthdeepgupta) [reply android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt:25]: these are under work

Comment (akanksha-einn) [reply android/MaterialDesignTest/app/src/main/java/com/testexample/materialdesigntest/data/interactor/implementation/ExaminationRepo.kt:25]: Ok

**Comment by akanksha-einn** [comment]
@nidhi-einn  I have reviewed the code.

**Comment by nidhi-einn** [comment]
@shresthdeepgupta have you tested this branch?

**Comment by shresthdeepgupta** [comment]
Unit testing is done but instrumental testing is pending

**Comment by nidhi-einn** [comment]
Okay I am merging on the basis of your unit testing

**Pull request merged**